src/bun.js/node/types.zig

```diff
@@ -2301,43 +2312,35 @@ pub const Path = struct {
         if (suffix_ptr) |_suffix_ptr| {
             // Supress exeption in zig. It does globalThis.vm().throwError() in JS land.
             validateString(globalObject, _suffix_ptr, "ext", .{}) catch {
-                return JSC.JSValue.jsUndefined();
+                // Returning .zero translates to a nullprt JSC.JSValue.
+                return .zero;
             };
         }
 
         const path_ptr = if (args_len > 0) args_ptr[0] else JSC.JSValue.jsUndefined();
         // Supress exeption in zig. It does globalThis.vm().throwError() in JS land.
         validateString(globalObject, path_ptr, "path", .{}) catch {
-            return JSC.JSValue.jsUndefined();
+            return .zero;
         };
 
         const pathZStr = path_ptr.getZigString(globalObject);
-        const len = pathZStr.len;
-        if (len == 0) return path_ptr;
+        if (pathZStr.len == 0) return path_ptr;
 
         var stack_fallback = std.heap.stackFallback(stack_fallback_size_small, JSC.getAllocator(globalObject));
         const allocator = stack_fallback.get();
 
         const pathZSlice = pathZStr.toSlice(allocator);
         defer pathZSlice.deinit();
 
+        var suffixZSlice: ?JSC.ZigString.Slice = null;
         if (suffix_ptr) |_suffix_ptr| {
             const suffixZStr = _suffix_ptr.getZigString(globalObject);
-            const suffixLen = suffixZStr.len;
-            if (suffixLen > 0 and suffixLen <= len) {
-                var suffixZSlice = suffixZStr.toSlice(allocator);
-                defer suffixZSlice.deinit();
-                return if (isWindows)
-                    toJSString(globalObject, basenameWindows(pathZSlice.slice(), suffixZSlice.slice()))
-                else
-                    toJSString(globalObject, basenamePosix(pathZSlice.slice(), suffixZSlice.slice()));
+            if (suffixZStr.len > 0 and suffixZStr.len <= pathZStr.len) {
+                suffixZSlice = suffixZStr.toSlice(allocator);
             }
         }
-
-        return if (isWindows)
-            toJSString(globalObject, basenameWindows(pathZSlice.slice(), null))
-        else
-            toJSString(globalObject, basenamePosix(pathZSlice.slice(), null));
+        defer if (suffixZSlice) |_s| _s.deinit();
+        return basenameJS_T(u8, globalObject, isWindows, pathZSlice.slice(), if (suffixZSlice) |_s| _s.slice() else null);
```

Comment:
Since we have the u8 and u16 versions, couldn’t we use the latin1 or utf16 string pointer directly and then not allocate in the first place?

Reply:
I think we should defer this to a follow-up PR at a later date

Reply:
@Jarred-Sumner Yes! This has come up before in review as a future todo too. We currently aren't using the u16 support that this PR makes way for so I imagine as we step into use things will get further refined.